src/wasm/wasm-validator.cpp

```diff
@@ -337,6 +342,82 @@ void FunctionValidator::noteLabelName(Name name) {
     "names in Binaryen IR must be unique - IR generators must ensure that");
 }
 
+void FunctionValidator::visitLoop(Loop* curr) {
+  if (curr->name.is()) {
+    noteLabelName(curr->name);
+    auto iter = breakInfos.find(curr->name);
+    assert(iter != breakInfos.end()); // we set it ourselves
+    auto& info = iter->second;
+    if (info.hasBeenSet()) {
+      shouldBeEqual(
+        info.arity, Index(0), curr, "breaks to a loop cannot pass a value");
+    }
+    breakInfos.erase(iter);
+  }
+  if (curr->type == none) {
+    shouldBeFalse(isConcreteType(curr->body->type),
+                  curr,
+                  "bad body for a loop that has no value");
+  }
+}
+
+// Validate correctness related to br_on_exn. br_on_exn is special: its argument
+// is an exnref value, and when it is taken, it pushes extracted values from
+// exnref, and if not, it re-pushes the exnref value onto the stack. To model
+// this value in binaryen AST, we calculate br_on_exn's type based on its
+// extracted event type, and the next instruction that takes re-pushed exnref
+// will get it from exnref.pop instruction instead. So within a block, if there
+// is a top-level br_on_exn (which is not dropped) and there is no exnref.pop
+// after that, that's a validation failure.
```

Comment:
I don't understand this modeling, I think. Why can't `br_on_exn` return a value normally, like `br_if`? (unlike `br_if` the branch value and the return value are different, but we can still use the returned value?) I remember we talked about this but I forgot the details, sorry.

Reply:
`br_if`'s AST node can contain the value it returns, like:
```
(br_if 0 (i32.const 3) (i32.const 1))
```
But for `br_on_exn` we don't know what value(s) will be returned within the AST code, because they will be extracted from the `exnref` argument.

The purpose of this validator was, even if there is a top-level `br_on_exn` which leaves a value on the stack that's not the same type as the block result type, if there is a `exnref.pop` after that, we shouldn't consider this as a validation failure.

Reply:
Talked with @kripken offline and decided to make `br_on_exn`s type as `exnref`, so we don't need this class anymore.